types/src/main/scala/aqua/types/IntersectTypes.scala

```diff
@@ -0,0 +1,66 @@
+package aqua.types
+
+import cats.Monoid
+import cats.data.NonEmptyMap
+
+object IntersectTypes extends Monoid[Type]:
+
+  override def empty: Type = TopType
+
+  def combineProducts(ap: ProductType, bp: ProductType): ProductType =
+    ProductType(
+      ap.toList.zip(bp.toList).map(combine)
+    )
+
+  override def combine(a: Type, b: Type): Type =
+    (a, b) match {
+      case (ap: ProductType, bp: ProductType) =>
+        combineProducts(ap, bp)
+
+      case (as: StructType, bs: StructType) =>
+        NonEmptyMap
+          .fromMap(as.fields.toSortedMap.flatMap { case (ak, at) =>
+            bs.fields.lookup(ak).map(bt => ak -> combine(at, bt))
+          })
+          .fold(empty)(fields => StructType(s"${as.name}_x_${bs.name}", fields))
+
+      case (aa: ArrowType, bb: ArrowType) =>
+        ArrowType(
+          UniteTypes.combineProducts(aa.domain, bb.domain),
+          combineProducts(aa.codomain, bb.codomain)
+        )
+
+      case (ac: OptionType, bc: BoxType) =>
+        OptionType(ac.element `∩` bc.element)
+
+      case (ac: BoxType, bc: OptionType) =>
+        OptionType(ac.element `∩` bc.element)
+
+      case (ac: ArrayType, bc: BoxType) =>
+        ArrayType(ac.element `∩` bc.element)
+      case (ac: BoxType, bc: ArrayType) =>
+        ArrayType(ac.element `∩` bc.element)
+      case (ac: StreamType, bc: StreamType) =>
+        StreamType(ac.element `∩` bc.element)
+
+      case (ScalarType.i64, ScalarType.u64) | (ScalarType.u64, ScalarType.i64) => ScalarType.u32
+      case (ScalarType.i64 | ScalarType.i32, ScalarType.u64 | ScalarType.u32) |
+          (ScalarType.u64 | ScalarType.u32, ScalarType.i64 | ScalarType.i32) =>
+        ScalarType.u16
+      case (
+            ScalarType.i64 | ScalarType.i16 | ScalarType.i32,
+            ScalarType.u64 | ScalarType.u16 | ScalarType.u32
+          ) | (
+            ScalarType.u64 | ScalarType.u16 | ScalarType.u32,
+            ScalarType.i64 | ScalarType.i16 | ScalarType.i32
+          ) =>
+        ScalarType.u8
+      case _ =>
```

Comment:
are you sure if this is a good way? if yes - add some comments that describes this wildcard